Use snowplow namespace (close #13)

matus-tomlein opened the pull request:
This PR actually does 2 things:

1. Adds a C++ namespace `snowplow` to all classes and other types in the tracker.
2. Formats the code using the [Google C++ Style Guide](https://google.github.io/styleguide/cppguide.html).
  * I like the idea of adopting an existing style guide that made some decisions and just making it consistent across the code base. It is based on the [Clang format style options](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) and there are more style guides to choose from (e.g., Visual Studio, LLVM, Mozilla).
  * I appreciate that this is a bit out of scope from the task of adding a namespace. I wanted to adopt a formatting while adding the namespace and it turned out a bigger change :) I can separate this out into another PR if that makes sense.

AlexBenny commented on examples/main.cpp:
```diff
@@ -1,18 +1,23 @@
+using std::string;
 
 void usage(char *program_name) {
   cout << "Usage: " << program_name << " [COLLECTOR_URI]" << endl;
 }
 
-int main(int argc, char** argv) {
+int main(int argc, char **argv) {
```
Nice, in this way it's more similar to the ObjC style guide

AlexBenny commented on performance/main.cpp:
```diff
@@ -2,47 +2,64 @@
+  double mute_emitter_and_mocked_session =
+      run_mute_emitter_and_mocked_session(db_name);
+  double mute_emitter_and_real_session =
+      run_mute_emitter_and_real_session(db_name);
```
I felt it was ok even before. 80 chars seems a quite strict rule for me :)
I think I'm in this group...
> Proponents of change argue that a wider line can make code more readable. The 80-column limit is an hidebound throwback to 1960s mainframes; modern equipment has wide screens that can easily show longer lines.
(https://google.github.io/styleguide/cppguide.html#Line_Length)

paulboocock replied:
FWIW, 80 char limits drive me nuts. I like a limit as very long lines can be hard to read but I think something around 120 is a much nicer modern number.

AlexBenny commented on performance/mock_emitter.hpp:
```diff
@@ -2,29 +2,35 @@
+ public:
+  MockEmitter(const string &db_name)
+      : Emitter("127.0.0.1:9090", Emitter::Method::POST,
+                Emitter::Protocol::HTTP, 52000, 52000, 500, db_name) {}
```
Personal taste, I'm not aligned with their approach to have multiple lines with multiple args.
In this case I would have preferred:

```cpp
MockEmitter(const string &db_name)
    : Emitter(
        "127.0.0.1:9090",
         Emitter::Method::POST,
         Emitter::Protocol::HTTP,
         52000,
         52000,
         500,
         db_name
     ) {}
```

AlexBenny commented on performance/run.hpp:
```diff
@@ -21,9 +23,9 @@ using std::string;
+double run_mocked_emitter_and_mocked_session(const string& db_name);
+double run_mocked_emitter_and_real_session(const string& db_name);
+double run_mute_emitter_and_mocked_session(const string& db_name);
+double run_mute_emitter_and_real_session(const string& db_name);
```
I think previously you moved the pointer near the variable name, like this:
```cpp
double run_mute_emitter_and_real_session(const string &db_name);
```
The style guide seems to be open to both options. Personally, I would prefer the pointer symbol near the var name as it would be the same for ObjC (IMO it makes sense from a grammatical point of view too), but open to what we prefer. Weird that they haven't set a clear rule for this.

AlexBenny commented on src/emitter.cpp:
```diff
@@ -2,28 +2,33 @@
-Emitter::Emitter(const string & uri, Method method, Protocol protocol, int send_limit, 
-  int byte_limit_post, int byte_limit_get, const string & db_name) : m_url(this->get_collector_url(uri, protocol, method)) {
+const int post_wrapper_bytes =
+    88;  // "schema":"iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-4","data":[]
```
Another point with I'm twisted by the rule.
In this case I wouldn't wrap with with the value in the newline. Also the comment would exceed in any case in the new line.

My preference would be for something like this:
```suggestion
// bytes length for schema
// (e.g.: "schema":"iglu:com.snowplowanalytics.snowplow/payload_data/jsonschema/1-0-4","data":[])
const int post_wrapper_bytes = 88;  
```
Even if I still prefer the original version.

AlexBenny commented on src/http_request_result.cpp:
```diff
@@ -42,10 +48,6 @@ int HttpRequestResult::get_http_response_code() {
-list<int> HttpRequestResult::get_row_ids() {
-  return this->m_row_ids;
-}
+list<int> HttpRequestResult::get_row_ids() { return this->m_row_ids; }
```
I wouldn't be against this in multiple rows.

AlexBenny commented on src/storage.hpp:
```diff
@@ -58,13 +64,14 @@ class Storage {
 
   void insert_payload(Payload payload);
   void insert_update_session(json session_data);
-  void select_all_event_rows(list<EventRow>* event_list);
-  void select_event_row_range(list<EventRow>* event_list, int range);
-  void select_all_session_rows(list<json>* session_list);
+  void select_all_event_rows(list<EventRow> *event_list);
+  void select_event_row_range(list<EventRow> *event_list, int range);
+  void select_all_session_rows(list<json> *session_list);
```
I think the position should be the same for both `&` and `*`.
I noticed that `&` was put near the type and the `*` near the var name.

AlexBenny reviewed (commented):
Following a style guide is a great idea.
Maybe it would have been better to split the changes in two issues but not a big problem.

I just disagree with the strict rule of the 80 chars.
IMO, in many places the code seems much harder to read. 
I would relax that rule. Also many of the changes in this PR are essentially a consequence of that rule.

The other point that I would reconsider is the position of the pointer symbol. I've seen it applied at the right or at the left in different places. We should decide where to put it. I'm favour of the right side as it's the same on ObjC.

Open to other opinions.

IMO we could also release this in the next version.

matus-tomlein committed:
Use snowplow namespace (close #13)

matus-tomlein committed:
Follow a code style guide (close #44)

matus-tomlein commented:
Thanks @AlexBenny and @paulboocock for comments! Made a few changes:

1. Split into two commits and made the second one refer to a new issue (I will rebase the PR when merging to keep the commits independent)
2. I change the style guide to LLVM instead of Google which reflects almost all your suggestions except for the line length
3. Added a `.clang-format` file which sets the LLVM style guide and adds a custom property to drop the requirement for max line length

AlexBenny reviewed (approved):
Thanks! Looks better to me now 👍 
Giant step forward
